mlir/lib/Transforms/lab4/simonyan_suren/SimonyanSurenFMAPass.cpp

```diff
@@ -0,0 +1,66 @@
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Tools/Plugins/PassPlugin.h"
+
+using namespace mlir;
+
+namespace {
+class SimonyanSurenFMAPass
+    : public PassWrapper<SimonyanSurenFMAPass, OperationPass<ModuleOp>> {
+private:
+  void handleMulOp(LLVM::FAddOp &addOp, LLVM::FMulOp &mulOp,
+                   Value &otherOperand) {
+    OpBuilder builder(addOp);
+    Value fma = builder.create<LLVM::FMAOp>(addOp.getLoc(), mulOp.getOperand(0),
+                                            mulOp.getOperand(1), otherOperand);
+    addOp.replaceAllUsesWith(fma);
+    if (mulOp.getOperand(0).hasOneUse()) {
+      mulOp.erase();
+    }
+    addOp.erase();
+  }
+
+public:
+  StringRef getArgument() const final { return "simonyan_suren_fma"; }
+  StringRef getDescription() const final {
+    return "Replaces add and multiply operations with a single instruction.";
+  }
+
+  void runOnOperation() override {
+    ModuleOp module = getOperation();
+    module.walk([&](Operation *operation) {
+      if (auto addOp = dyn_cast<LLVM::FAddOp>(operation)) {
+        Value addLHS = addOp.getOperand(0);
+        Value addRHS = addOp.getOperand(1);
+
+        if (auto mulOpLHS = addLHS.getDefiningOp<LLVM::FMulOp>()) {
+          handleMulOp(addOp, mulOpLHS, addRHS);
+        } else if (auto mulOpRHS = addRHS.getDefiningOp<LLVM::FMulOp>()) {
+          handleMulOp(addOp, mulOpRHS, addLHS);
+        }
+      }
+    });
+
+    module.walk([&](Operation *operation) {
+      if (auto mulOp = dyn_cast<LLVM::FMulOp>(operation)) {
+        if (mulOp.use_empty()) {
+          mulOp.erase();
+        }
+      }
+    });
+  }
+};
+} // namespace
+
+MLIR_DECLARE_EXPLICIT_TYPE_ID(SimonyanSurenFMAPass)
+MLIR_DEFINE_EXPLICIT_TYPE_ID(SimonyanSurenFMAPass)
+
+PassPluginLibraryInfo getSimonyanSurenFMAPassPluginInfo() {
+  return {MLIR_PLUGIN_API_VERSION, "kulaev_zhenya_fma", LLVM_VERSION_STRING,
```

Comment:
It looks suspicious

Reply:
Sorry, I forgot to change it, we did it together with a colleague